### PR TITLE
Feat: Avd install for codespaces

### PR DESCRIPTION
**Pull request by ankudinov**
## Change Summary

This PR add support for AVD installation inside Codespaces.
Currently mount fails due to: https://github.com/orgs/community/discussions/85658
As a workaround, `dev` image `entrypoint.sh` will get additional logic to install AVD directly from the workspace.

## Component(s) name

`containers`

## How to test
Once AVD dev image will be re-build after merge, start Github Codespace on any PR and check:
- ansible --version
- ansible-galaxy collection list

It should be also possible to "unpack" AVD examples and build configs in every example directory.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -1,8 +0,0 @@
-ARG VARIANT="3.8-v2.0"
```
Comment (gmuloc) [inline .devcontainer/Dockerfile:1]: why is this removed in this PR - is it related?

File: containers/dev/.devcontainer/entrypoint.sh [hunk 2]
```diff
@@ -24,6 +24,20 @@ ansible --version &> /dev/null ||  if ! [ -z "${AVD_GITHUB_REPO}" ] && ! [ -z "$
...
+AVD_DEV_REQ_FILE_ALT="${CONTAINER_WORKSPACE}/ansible_collections/arista/avd/requirements-dev.txt"
+ansible --version &> /dev/null ||  if [ -f $AVD_DEV_REQ_FILE_ALT ]; then
+  sudo chown -R ${USERNAME} ${HOME}/.ansible
+  ANSIBLE_CORE_VERSION=$(cat ${AVD_DEV_REQ_FILE_ALT}| grep ansible-core)
+  pip3 install "${ANSIBLE_CORE_VERSION}"
+  ansible-galaxy collection install ${CONTAINER_WORKSPACE}/ansible_collections/arista/avd/
+  ansible-galaxy collection install -r ${CONTAINER_WORKSPACE}/ansible_collections/arista/avd/collections.yml
+  pip3 install -r ${CONTAINER_WORKSPACE}/ansible_collections/arista/avd/requirements.txt
+  pip3 install -r ${CONTAINER_WORKSPACE}/ansible_collections/arista/avd/requirements-dev.txt
+fi
```
Comment (gmuloc) [inline containers/dev/.devcontainer/entrypoint.sh]: This is almost the same block as line 6 to 14 - is it not possible to just make some logic to decide what the `ROOT` should be and then based on this use `ROOT` everwhere?

like (not proper bash ;)) `ROOT = CONTAINER_WORKSPACE or ${HOME}/.ansible/collections`

File: containers/dev/.devcontainer/entrypoint.sh [hunk 3]
```diff
@@ -24,6 +24,20 @@ ansible --version &> /dev/null ||  if ! [ -z "${AVD_GITHUB_REPO}" ] && ! [ -z "$
...
 fi
 
+# In some cases AVD can not be correctly mounted, for ex. when running dev container as Codespace
+# In that case if collection is available in the container workspace, it will be installed from there
+CONTAINER_WORKSPACE=$(git rev-parse --show-toplevel)
+AVD_DEV_REQ_FILE_ALT="${CONTAINER_WORKSPACE}/ansible_collections/arista/avd/requirements-dev.txt"
+ansible --version &> /dev/null ||  if [ -f $AVD_DEV_REQ_FILE_ALT ]; then
+  sudo chown -R ${USERNAME} ${HOME}/.ansible
+  ANSIBLE_CORE_VERSION=$(cat ${AVD_DEV_REQ_FILE_ALT}| grep ansible-core)
+  pip3 install "${ANSIBLE_CORE_VERSION}"
```
Comment (gmuloc) [inline containers/dev/.devcontainer/entrypoint.sh:26]: does it need to be `pip3` everywhere because the base container still has python2 or is just some sort of clarity?

File: .devcontainer/Dockerfile (hunk 1 above)
Comment (ankudinov) [reply .devcontainer/Dockerfile:1]: not related. It's not in use any more since devcontainer.json update. But file wasn't removed in the past.

File: containers/dev/.devcontainer/entrypoint.sh (hunk 2 above)
Comment (ankudinov) [reply containers/dev/.devcontainer/entrypoint.sh]: I'd prefer to keep it this way. More readable and works. Any bash optimisation - separate PR.

File: containers/dev/.devcontainer/entrypoint.sh (hunk 3 above)
Comment (ankudinov) [reply containers/dev/.devcontainer/entrypoint.sh:26]: everything is Python3. All AVD dev containers are based on python-slim with 3.9, 3.10, 3.11 Python.

File: containers/dev/.devcontainer/entrypoint.sh (hunk 2 above)
Comment (gmuloc) [reply containers/dev/.devcontainer/entrypoint.sh]: this opens the field for more mistake in future PR to change both blocks. looks like the order is important in which we try to install ansible? Is it documented somewhere

Comment (ClausHolbechArista) [reply containers/dev/.devcontainer/entrypoint.sh]: please combine the two pip installs into a single command with two -r. By doing that you will resolve any conflicts in one step.

**Review by ClausHolbechArista — APPROVED** [review]

**Comment by ankudinov** [comment]
Due to the comments, `entrypoint.sh` was changed significantly and requires full testing.
Please proceed with the following steps to prepare local dev container build:
- clone AVD repository
- create a new branch for testing purposes only (it can be deleted once testing is finished)
- add following files to `.devcontainer/` directory:

  - `Dockerfile`

    ```Dockerfile
    ARG FROM_IMAGE
    ARG FROM_VARIANT

    FROM ${FROM_IMAGE}:${FROM_VARIANT}

    ARG USERNAME

    # Add entrypoint script.
    COPY ./entrypoint.sh /bin/entrypoint.sh
    RUN chmod +x /bin/entrypoint.sh
    # Use ENTRYPOINT instead of CMD to ensure that entryscript is always executed.
    ENTRYPOINT [ "/bin/entrypoint.sh" ]

    # Switch user to given USERNAME otherwise Ansible will be installed as root.
    USER ${USERNAME}
    ENV PATH=$PATH:/home/${USERNAME}/.local/bin
    ```

  - `devcontainer.json`

    ```json
    // For format details, see https://aka.ms/devcontainer.json. For config options, see the
    // README at: https://github.com/devcontainers/templates/tree/main/src/python.
    {
        "build": {
            "dockerfile": "./Dockerfile",
            "context": ".",
            "args": {
                "FROM_IMAGE": "ghcr.io/aristanetworks/ansible-avd/base",
                "FROM_VARIANT": "python3.11",
                "USERNAME": "avd"
            }
        },
        "features": {
            "ghcr.io/devcontainers/features/docker-in-docker:2": {
                "version": "latest",
                "dockerDashComposeVersion": "none"
            },
            // Add sshd to support gh cli codespace cp.
            "ghcr.io/devcontainers/features/sshd:1": {
                "version": "latest"
            }
        },
        "customizations": {
            "vscode": {
                "extensions": [
                    // Git essentials.
                    "github.vscode-pull-request-github",
                    // Python.
                    "ms-python.python",
                    "ms-python.vscode-pylance",
                    // Errors and highlighters.
                    "mechatroner.rainbow-csv",
                    "oderwat.indent-rainbow",
                    // Excalidraw.
                    "pomdtr.excalidraw-editor"
                ]
            }
        }
    }
    ```

- copy `entrypoint.sh` from this PR to the root of AVD repo or `.devcontainer/` and make it executable.

Tests:

1. Check that container can be created without github repo defined and without mounts.

  - build devcontainer. It's quite long and boring. Good chance to drink a coffee, but a small one.
  - run `entrypoint.sh`
  - verify that there are no errors and if following commands provide meaninful results:
    
    - `ansible --version`
    - `ansible-galaxy collection list`

2. Check that collection can be installed from github branch

    - add following to devcontainer.json

      ```json
      "containerEnv": {
        "AVD_GITHUB_REPO": "aristanetworks/ansible-avd",
        "AVD_BRANCH_NAME": "devel"
      },
      ```
    
    - build container, run `entrypoint.sh`, verify

3. Check that collection can be installed from a mount

    - remove env variables from `devcontainer.json`
    - add following settings

        ```json
        "mounts": [
            "source=${localWorkspaceFolder}/ansible_collections/arista/avd,target=/home/avd/.ansible/collections/ansible_collections/arista/avd,type=bind"
        ],
        ```
    
    - build container, run `entrypoint.sh`, verify

**Review by gmuloc — APPROVED** [review]
LGTM
